dev/com.ibm.ws.logging/src/com/ibm/ws/logging/data/GenericData.java

```diff
@@ -10,68 +10,94 @@
  *******************************************************************************/
 package com.ibm.ws.logging.data;
 
-import java.util.ArrayList;
-
 public class GenericData {
 
-    private final ArrayList<KeyValuePair> pairs;
+    private final static int DEFAULT_SIZE = 16;
+
+    private KeyValuePair[] pairs;
 
     private String sourceType;
 
     private String jsonMessage = null;
 
+    private Integer lastIndex = -1;
+
     public GenericData() {
-        pairs = new ArrayList<KeyValuePair>();
+        pairs = new KeyValuePair[DEFAULT_SIZE];
     }
 
     public GenericData(int size) {
-        pairs = new ArrayList<KeyValuePair>(size);
-    }
-
-    public void modifyPair(int index, String key, String value) {
-        KeyValueStringPair kvp = new KeyValueStringPair(key, value);
-        pairs.set(index, kvp);
+        pairs = new KeyValuePair[size];
     }
 
     public void setPair(int index, String key, String value) {
         KeyValueStringPair kvp = new KeyValueStringPair(key, value);
-        pairs.add(index, kvp);
+        ensureCapacityAndSetPair(index, kvp);
+        setLastIndex(index);
```

Comment:
code to set the lastIndex (currently in setLastIndex) should be inside ensureCapacityAndSetPair since every time you want to setLastIndex you want to potentially set lastIndex.  This would also be consistent with how you're handling ensureCapacityAndAddPair.